tests/test_pgsql_parse.py

```diff
@@ -0,0 +1,616 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2012-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import unittest
+from edb.testbase import server as tb
+
+from edb.pgsql import parser
+from edb.pgsql import codegen
+
+
+def parse_and_gen(sql: str) -> str:
+    ast = parser.parse(sql)
+    sql_stmts = [codegen.generate_source(stmt, pretty=False) for stmt in ast]
+    sql = "; ".join(sql_stmts)
+    return sql.replace("  ", " ").replace("( (", "((").replace(") )", "))")
+
+
+class TestEdgeQLSelect(tb.TestCase):
```

Comment:
I forgot this, but we have a framework for writing tests where input and output can be specified as strings. An example is here: https://github.com/edgedb/edgedb/blob/master/tests/test_edgeql_ir_card_inference.py#L31

Your call if you want to try to use it.

I think here it might actually make things worse, since you properly handle whitespace in your tests. I think that for the parse/codegen tests that we have for the edgeql parser, we actually do the comparison after stripping away all whitespace!